### ISSUE TITLE
How to use GritLM as a generative reranker?

### ISSUE THREAD

**Issue by orionw** [ContextualAI/gritlm#25]
Hi @Muennighoff, thanks again for your great work here!

I am trying to use GritLM as a reranker, but it wasn't doing too great so I thought I'd ask for the recommended way to use it for reranking. I implemented it using the same approach as MonoT5, where the model compares the softmax scores of `true` vs `false` given a prompt asking to do so.  I used the following prompt:

```python
template = """<s>[INST] You are an expert Google searcher, whose job is to determine if the following document is relevant to the query (true/false).
Query: {query}
Document: {text}
Relevant (either "true" or "false"): [/INST]"""
```

However, GritLM's generative ability is not designed for that token comparison - would you suggest any other approaches? Perhaps generating the token would be preferable or perhaps I am using the wrong template?

FWIW, the average p-MRR score on FollowIR for following instructions using this approach was 7.9, much higher than its embedding p-MRR score of 1.1.  But the overall reranking nDCG averaged to 5.8 and I think it's likely GritLM can do better.

**Comment by Muennighoff** [comment]
Very cool! The format should be different using <|user|> etc., see [here](https://github.com/ContextualAI/gritlm/blob/01b2ed4799f09df2d9d41a9870176606b32568aa/scripts/AbsTaskRetrieval.py#L96) which is what we used for the reranking in the paper. It's also a slightly different technique, but I think your technique should work too with the right template. Let me know if that helps and if not I'll look deeper!

**Comment by orionw** [comment]
I added the referenced implementation and then realized that this style of reranking (taking in all passages at once) doesn't generalize to longer passages well, in terms of number of passages or their word count.  I can only fit < 5 passages at a time to rerank with a 2k context limit, which doesn't allow me to rerank the top 1000.

I think there could be some sort of bubble sort / list-wise approach, but nothing trivial came to my mind - unless you had a suggestion? 

I did try the same style prompt with the MonoT5 approach (below) and got better results from the Mistral type prompt, for an average of 10.2 nDCG and 7.5 p-MRR - still seems like it's an unfair comparison though since it wasn't trained on that style reranking, but pretty good results zero-shot.

New prompt with MonoT5 style reranking:
```
self.template = "<|user|>\nRank the passage based on its relevance to the search query (true/false) {query}.\n\n{text}\n\n" \
                   "Search Query: {query}.\n\n" \
                   "Determine the passage's relevance to the search query." \
                   "Only respond with true or false, do not say any other word or explain.\n<|assistant|>\n"
```

**Comment by Muennighoff** [comment]
> I added the referenced implementation and then realized that this style of reranking (taking in all passages at once) doesn't generalize to longer passages well, in terms of number of passages or their word count. I can only fit < 5 passages at a time to rerank with a 2k context limit, which doesn't allow me to rerank the top 1000.
> 
> I think there could be some sort of bubble sort / list-wise approach, but nothing trivial came to my mind - unless you had a suggestion?
> 
> I did try the same style prompt with the MonoT5 approach (below) and got better results from the Mistral type prompt, for an average of 10.2 nDCG and 7.5 p-MRR - still seems like it's an unfair comparison though since it wasn't trained on that style reranking, but pretty good results zero-shot.
> 
> New prompt with MonoT5 style reranking:
> 
> ```
> self.template = "<|user|>\nRank the passage based on its relevance to the search query (true/false) {query}.\n\n{text}\n\n" \
>                    "Search Query: {query}.\n\n" \
>                    "Determine the passage's relevance to the search query." \
>                    "Only respond with true or false, do not say any other word or explain.\n<|assistant|>\n"
> ```

Very cool!

I remember that @hongjin-su did something to solve the context problem - could you elaborate @hongjin-su? (also you can go much bigger than 2K - 4096 should also work very well and maybe even 32768).

Also is it on purpose that the query appears twice in your prompt `{query}`?

**Comment by orionw** [comment]
> I remember that @hongjin-su did something to solve the context problem - could you elaborate @hongjin-su? 

Would definitely be interested in hearing it!

> (also you can go much bigger than 2K - 4096 should also work very well and maybe even 32768).

4096 would allow for reranking 10 ish longer documents of size 400, but that's not enough to do FollowIR which will require reranking 1000...  Even 32768 would only get to ~80 docs.  My guess is that current LMs aren't that good at long-context stuff enough to handle that scale of reranking in the prompt.  It would actually be an interesting experiment though with a model that can do 500k+ context length (say Gemini 1.5 Pro). Probably a paper/evaluation task there for long context models...

> Also is it on purpose that the query appears twice in your prompt {query}?

Yes, it was in your [initial prompt](https://github.com/ContextualAI/gritlm/blob/01b2ed4799f09df2d9d41a9870176606b32568aa/scripts/AbsTaskRetrieval.py#L96) - I tried to keep as much of the original formatting as possible.

**Comment by hongjin-su** [comment]
Thanks for the question!

To incorporate long context, I do pairwise comparisons between the query and each of top-10 doc. Specifically, I ask LM to output yes if query and doc are related, and no otherwise. The probability for yes is used as the criteria for reranking. https://github.com/ContextualAI/gritlm/pull/26/commits/ccf8f6769dde9561dde61ce8df09683c517f5c42

Hope this helps!
